File: Ruby/Civitas/lib/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "titulo_propiedad"

module Civitas
  class JugadorEspeculador < Jugador
    def initialize(jugador, fianza)
      nuevo_especulador(jugador)
      @@FactorEspeculador=2
      @fianza=fianza
      @especulador=true
    end
    
    attr_reader :FactorEspeculador
    attr_accessor :fianza, :especulador
    
    def nuevo_especulador(jugador)
      super.copia(jugador)
      
      super.propiedades.each do |propiedad|
        propiedad.actualizar_propietario_por_conversion(self)
      end
      
    end
    
    # Método sobrescrito
    def encarcelar(num_casilla__carcel)
      aux=false
      
      if(super.encarcelado==false)
        unless(super.tiene_Salvoconducto)
          if(fianza>=super.saldo)
            aux=true
          else
            super.paga(fianza)
            Diario.instance.ocurre_evento("El jugador #{super.nombre} se ha librado de la cárcel pagando una fianza. \n")
          end
        else
          aux=false
          super.perder_salvoconducto
          Diario.instance.ocurre_evento("El jugador #{super.nombre} se ha librado de la cárcel usando el salvoconducto. \n")
        end
      end
      
      return aux
    end
    
    # Método sobrescrito
    def casas_max
      return super.Casas_max*FactorEspeculador
    end
    
    # Método sobrescrito
    def hoteles_max
      return super.Hoteles_max*FactorEspeculador
    end
    
    # Método sobrescrito
    def paga_impuesto(cantidad)
      aux=false
      
      unless(encarcelado)
        impuesto=cantidad/FactorEspeculador
        aux=paga(impuesto)
        Diario.instance.ocurre_evento("El jugador")
      end
      return aux
    end
    
    def to_s
      return "\nNombre: #{super.nombre}
              \nEspeculador: #{especulador}
              \nCasilla actual: #{super.num_casilla_actual}
              \nEncarcelado: #{super.encarcelado}
              \nPuede comprar: #{super.puede_comprar}
              \nSaldo: #{super.saldo}
              \nNumero de propiedades: #{super.get_propiedades}
              \nPropiedades: #{super.propiedades}"
    end
  end
end

File: Java/Civitas/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author arturo
 */
public class JugadorEspeculador extends Jugador{
    private static int FactorEspeculador=2;
    private float fianza;
    private boolean especulador;
    
    JugadorEspeculador(Jugador jugador, float fianza){
        super(jugador);
        this.fianza=fianza;
        especulador=true;
        
        for(int i=0; i<super.getPropiedades().size(); i++){
            super.getPropiedades().get(i).actualizaPropietarioPorConversion(this);
        }
    }
    
    @Override
    boolean encarcelar(int numCasillaCarcel){
        boolean aux=false;
        
        if(super.encarcelado==false){
            if(super.tieneSalvoconducto()==false){
                if(fianza>=super.getSaldo()){
                aux=true;
            }else{
                super.paga(fianza);
                Diario.getInstance().ocurreEvento("El jugador "+super.getNombre()+" Se ha librado de la cárcel pagando una fianza\n");
            }
            }else{
                aux=false;
                super.perderSalvoconducto();
                Diario.getInstance().ocurreEvento("El jugador "+super.getNombre()+" se libra de la cárcel con el salvoconducto\n");
            }
        }
    return aux;        
    }
    
    @Override
    int getCasasMax(){
        return super.getCasasMax()*FactorEspeculador;
    }
    
    @Override 
    int getHotelesMax(){
        return super.getHotelesMax()*FactorEspeculador;
    }
    
    @Override 
    boolean pagaImpuesto(float cantidad){
        boolean aux=false;
        
        if(encarcelado==false){
            float impuesto=cantidad/2;
            aux=paga(impuesto);
            Diario.getInstance().ocurreEvento("El Jugador "+super.getNombre()+" paga "+impuesto+" €\n");
        }
    return aux;
    }
    
    @Override
    public String toString(){
        return "\n"+super.getNombre()+"\nEncarcelado: "+super.encarcelado
                +"\nPuede comprar: "+super.getPuedeComprar()+"\nSaldo: "+super.getSaldo()+"\nNumero de propiedades: "+
                super.getPropiedades().size()+"\nPropiedades: "+super.getPropiedades()+" "+"\nCasilla actual: "+super.getNumCasillaActual()+
                "\nEspeculador: "+especulador+"\n-----------------------------------------";
        
    }
}
